### PR TITLE
Dynamically set prod and network code 

### PR DESCRIPTION
**Pull request by Lindani**
in prep_hotsocket_data function

### CODE REVIEW

**Comment by Lindani** [comment]
@gsvr : Please review

File: recharges/tests.py [hunk 1]
```diff
@@ -210,6 +210,70 @@ def test_update_recharge_status_hotsocket_ref(self):
         recharge = Recharge.objects.get(id=recharge_id)
         self.assertEqual(recharge.hotsocket_ref, 555)
 
+    def test_normalize_msisdn(self):
+        # Setup
+        self.make_account()
+
+        recharge_id = self.make_recharge(msisdn="+2772")
+        # Execute
+        result = hotsocket_get_airtime.normalize_msisdn(recharge_id,
```
Comment (gsvr) [inline recharges/tests.py]: You should pass the msisdn in here, not the recharge_id.

File: recharges/tests.py [hunk 2]
```diff
@@ -210,6 +210,70 @@ def test_update_recharge_status_hotsocket_ref(self):
...
+        self.assertEqual(result, "+27724455545")
+
+        # Setup
+        recharge_id = self.make_recharge(msisdn="27724455545")
+        # Execute
+        result = hotsocket_get_airtime.normalize_msisdn(recharge_id,
+                                                        country_code='27')
+        # Check
+        self.assertEqual(result, "+27724455545")
+
```
Comment (gsvr) [inline recharges/tests.py]: Let's also make the function strip out any spaces in the msisdn.

**Comment by Lindani** [comment]
@gsvr : Please re-review

File: recharges/tests.py [hunk 3]
```diff
@@ -321,20 +388,21 @@ def test_hotsocket_get_airtime_good(self):
             json.dumps(expected_response_good),
             status=200, content_type='application/json')
 
-        recharge_id = self.make_recharge(msisdn="+277244555", status=0)
+        recharge_id = self.make_recharge(msisdn="+27 7244555", status=0)
```
Comment (gsvr) [inline recharges/tests.py]: It bugs me a little bit that this number is not the correct length - one character short.

**Comment by gsvr** [comment]
@imsickofmaps Please review.

File: recharges/tasks.py [hunk 4]
```diff
@@ -108,6 +108,38 @@ class Hotsocket_Get_Airtime(Task):
...
+        if msisdn.startswith('0'):
+            return '+' + country_code + msisdn[1:]
+        if msisdn.startswith('+'):
+            return msisdn
+        if msisdn.startswith(country_code):
+            return '+' + msisdn
+        return msisdn
+
+    def look_up_mobile_operator(self, msisdn):
+        mtn = ['+2783', '+2773']
```
Comment (imsickofmaps) [inline recharges/tasks.py]: These lists are incomplete. Please review against: https://en.wikipedia.org/wiki/Telephone_numbers_in_South_Africa

Comment (imsickofmaps) [inline recharges/tasks.py]: Also, some are shorter prefixes than others - so you'll need to allow for that in this code, try a short match then a longer match.

File: recharges/tasks.py [hunk 5]
```diff
@@ -108,6 +108,38 @@ class Hotsocket_Get_Airtime(Task):
     """
     name = "recharges.tasks.hotsocket_get_airtime"
 
+    def normalize_msisdn(self, msisdn, country_code='',):
```
Comment (imsickofmaps) [inline recharges/tasks.py]: These two functions are lacking comments. Also, I think they're useful enough that we should keep them outside the class for reuse.

**Comment by imsickofmaps** [comment]
I'm done for now

**Comment by Lindani** [comment]
@gsvr : Please re-review

File: recharges/tasks.py [hunk 6]
```diff
@@ -144,11 +193,17 @@ def run(self, recharge_id, **kwargs):
...
-
         if status == 0:
             recharge.status = 1
             recharge.save()
 
+            recharge.msisdn = normalize_msisdn(recharge.msisdn)
+            recharge.save()
+
+            recharge.network_code = look_up_mobile_operator(recharge.
+                                                            msisdn)
```
Comment (gsvr) [inline recharges/tasks.py]: This can fit on one line

File: recharges/tests.py [hunk 7]
```diff
@@ -127,13 +128,21 @@ def test_prep_hotsocket_data(self):
...
         # Check
-        self.assertEqual(hotsocket_data["recipient_msisdn"], "+27123")
+        self.assertEqual(hotsocket_data["recipient_msisdn"], "+2782003453")
         self.assertEqual(hotsocket_data["token"], '1234')
+        self.assertEqual(hotsocket_data["product_code"], 'AIRTIME')
+        self.assertEqual(hotsocket_data["network_code"], 'VOD')
+
         self.assertEqual(hotsocket_data["reference"], recharge_id + 10000)
 
+        # check product_code and network
```
Comment (gsvr) [inline recharges/tests.py]: rogue comment

File: recharges/tests.py [hunk 8]
```diff
@@ -26,7 +27,7 @@ def make_account(self, token='1234'):
         account = Account.objects.create(token=token)
         return account.id
 
-    def make_recharge(self, amount=100.00, msisdn="+27123", status=0):
+    def make_recharge(self, amount=100.00, msisdn="+2782003453", status=0):
```
Comment (gsvr) [inline recharges/tests.py]: This number is also short.

File: recharges/tests.py [hunk 9]
```diff
@@ -321,20 +399,21 @@ def test_hotsocket_get_airtime_good(self):
...
             status=200, content_type='application/json')
 
-        recharge_id = self.make_recharge(msisdn="+277244555", status=0)
+        recharge_id = self.make_recharge(msisdn="+27 71145557", status=0)
         # Execute
         result = hotsocket_get_airtime.delay(recharge_id)
         # Check
         self.assertEqual(result.get(),
-                         "Recharge for +277244555: Queued at Hotsocket #4487")
+                         "Recharge for +2771145557: Queued at Hotsocket #4487")
```
Comment (gsvr) [inline recharges/tests.py]: Sorry, it's still short - my bad.

**Comment by gsvr** [comment]
@imsickofmaps Please re-review.

File: recharges/tasks.py [hunk 10]
```diff
@@ -120,10 +180,10 @@ def prep_hotsocket_data(self, recharge_id):
             'as_json': True,
             'token': get_token(),
             'recipient_msisdn': recharge.msisdn,
-            'product_code': 'DATA',  # TODO 37: to dynamically set product code
-            'network_code': 'VOD',  # TODO 37: to dynamically set network code
+            'product_code': recharge.product_code,
+            'network_code': recharge.network_code,
             'denomination': recharge.amount,
-            'reference': recharge.id + settings.HOTSOCKET_REFBASE
+            'reference': recharge_id + settings.HOTSOCKET_REFBASE
```
Comment (imsickofmaps) [inline recharges/tasks.py]: you need cast `settings.HOTSOCKET_REFBASE` as an integer because environment vars are strings

Comment (imsickofmaps) [inline recharges/tasks.py]: or do that in settings

File: recharges/tasks.py [hunk 11]
```diff
@@ -144,11 +205,16 @@ def run(self, recharge_id, **kwargs):
...
         status = recharge.status
-
         if status == 0:
             recharge.status = 1
             recharge.save()
 
+            recharge.msisdn = normalize_msisdn(recharge.msisdn)
+            recharge.save()
+
+            recharge.network_code = look_up_mobile_operator(recharge.msisdn)
```
Comment (imsickofmaps) [inline recharges/tasks.py]: What if this returns `False`?

File: recharges/tests.py [hunk 12]
```diff
@@ -321,20 +397,21 @@ def test_hotsocket_get_airtime_good(self):
...
         recharge = Recharge.objects.get(id=recharge_id)
-
         self.assertEqual(recharge.status, 1)
         self.assertEqual(recharge.hotsocket_ref, 4487)
         """tests for the correct URL request"""
         self.assertEqual(len(responses.calls), 1)
         self.assertEqual(responses.calls[0].request.url,
                          "http://test-hotsocket/recharge")
+        self.assertEqual(recharge.msisdn, '+27711455657')
+        self.assertEqual(recharge.network_code, 'VOD')
```
Comment (imsickofmaps) [inline recharges/tests.py]: Need a test for a submission that wasn't for a valid network to make sure it doesn't hit hotsocket

**Comment by imsickofmaps** [comment]
@Lindani @gsvr I'm done again

File: gopherairtime/settings.py [hunk 13]
```diff
@@ -171,7 +171,7 @@
 HOTSOCKET_API_ENDPOINT = 'http://api.hotsocket.co.za:8080/test'
 HOTSOCKET_API_USERNAME = 'Replaceme_username'
 HOTSOCKET_API_PASSWORD = 'Replaceme_password'
-HOTSOCKET_REFBASE = 10000
+HOTSOCKET_REFBASE = "10000"
```
Comment (imsickofmaps) [inline gopherairtime/settings.py]: needs the env lookup too

**Comment by imsickofmaps** [comment]
:+1: after env change

**Pull request merged**